rust/parser/mod.rs

```diff
@@ -49,279 +46,381 @@ use crate::{
         TypeQLUndefine, TypeQLUpdate,
     },
 };
-use antlr_rust::{
-    token::Token,
-    tree::{ParseTree, TerminalNode as ANTLRTerminalNode},
-};
 use chrono::{NaiveDate, NaiveDateTime};
-use std::rc::Rc;
+use pest::Parser;
+use pest_derive::Parser;
 
-// keep star import to not expose generated code
-use typeql_grammar::typeqlrustparser::*;
+#[derive(Parser)]
+#[grammar = "parser/typeql.pest"]
+struct TypeQLParser;
 
-type TerminalNode<'a> = ANTLRTerminalNode<'a, TypeQLRustParserContextType>;
+type SyntaxTree<'a> = pest::iterators::Pair<'a, Rule>;
+type Children<'a> = pest::iterators::Pairs<'a, Rule>;
 
-fn get_string(string: Rc<TerminalNode>) -> String {
-    unquote(&string.get_text())
+trait TreeChildren {
+    type Items;
+    fn into_children(self) -> Self::Items;
 }
 
-fn get_regex(string: Rc<TerminalNode>) -> String {
-    unescape_regex(&unquote(&string.get_text()))
+impl<'a> TreeChildren for SyntaxTree<'a> {
+    type Items = Children<'a>;
+    fn into_children(self) -> Self::Items {
+        self.into_inner()
+    }
 }
 
-fn get_long(long: Rc<TerminalNode>) -> i64 {
-    long.get_text().parse().expect(&ILLEGAL_GRAMMAR.format(&[long.get_text().as_str()]).message)
+trait RuleIterator {
+    type Item;
+
+    fn skip_expected(&mut self, rule: Rule) -> &mut Self;
+    fn consume_expected(&mut self, rule: Rule) -> Self::Item;
+    fn peek_rule(&mut self) -> Option<Rule>;
+    fn consume_if_matches(&mut self, rule: Rule) -> Option<Self::Item>;
+    fn consume_any(&mut self) -> Self::Item;
+    fn try_consume_any(&mut self) -> Option<Self::Item>;
 }
 
-fn get_double(double: Rc<TerminalNode>) -> f64 {
-    double.get_text().parse().expect(&ILLEGAL_GRAMMAR.format(&[double.get_text().as_str()]).message)
+impl<'a, T: Iterator<Item = SyntaxTree<'a>> + Clone> RuleIterator for T {
+    type Item = SyntaxTree<'a>;
+
+    fn skip_expected(&mut self, rule: Rule) -> &mut Self {
+        self.consume_expected(rule);
+        self
+    }
+
+    fn consume_expected(&mut self, _rule: Rule) -> Self::Item {
+        let next = self.consume_any();
+        assert_eq!(next.as_rule(), _rule);
+        next
+    }
+
+    fn peek_rule(&mut self) -> Option<Rule> {
+        self.clone().peekable().peek().map(|tree| tree.as_rule())
+    }
```

Comment:
`peeakable()` takes ownership, so we give it a clone of the iterator instead.

Reply:
is this performant if it happens a lot during parsing?

Reply:
Iterators don't store a lot of state (basically just a pointer and maybe a closure), so in general cloning them is fine. In this specific instance, the calls don't even show up in the profiler, so we should be in the clear.